packages/@aws-cdk/aws-appsync/lib/graphqlapi.ts

```diff
@@ -433,6 +467,174 @@ export class GraphQLApi extends Construct {
+          if (resource.custom) {
+            resourceTypes = `${resource.custom}`;
+          } else if (resource.type && resource.field) {
+            resourceTypes = `types/${resource.type}/fields/${resource.field}`;
+          } else if (resource.type) {
+            resourceTypes = `types/${resource.type}/*`;
+          } else {
+            resourceTypes = '*';
+          }
+          return Stack.of(this).formatArn({
+            service: 'appsync',
+            resource: `apis/${this.apiId}`,
+            sep: '/',
+            resourceName: `${resourceTypes}`,
+          });
+        },
+      }))],
+      scope: this,
+    });
+
+  }
+
+  /**
+   * Adds an IAM policy statement for full access to this GraphQLApi to an IAM
+   * principal's policy.
+   *
+   * @param grantee The principal (no-op if undefined)
+   */
+  public grantFullAccess(grantee: IGrantable): Grant {
+    return Grant.addToPrincipal({
+      grantee,
+      actions: ['appsync:graphql'],
+      resourceArns: [
+        Stack.of(this).formatArn({
+          service: 'appsync',
+          resource: `apis/${this.apiId}`,
+          sep: '/',
+          resourceName: '*',
+        }),
+      ],
+      scope: this,
+    });
+  }
+
+  /**
+   * Adds an IAM policy statement for Mutation access to this GraphQLApi to an IAM
+   * principal's policy.
+   *
+   * @param grantee The principal (no-op if undefined)
+   * @param fields The fields to grant access to that are Mutations (leave blank for all)
+   */
+  public grantMutation(grantee: IGrantable, fields?: [string]): Grant {
+    return this.grantType(grantee, 'Mutation', fields);
+  }
+
+  /**
+   * Adds an IAM policy statement for Query access to this GraphQLApi to an IAM
+   * principal's policy.
+   *
+   * @param grantee The principal (no-op if undefined)
+   * @param fields The fields to grant access to that are Queries (leave blank for all)
+   */
+  public grantQuery(grantee: IGrantable, fields?: [string]): Grant {
+    return this.grantType(grantee, 'Query', fields);
+  }
+
+  /**
+   * Adds an IAM policy statement for Subscription access to this GraphQLApi to an IAM
+   * principal's policy.
+   *
+   * @param grantee The principal (no-op if undefined)
+   * @param fields The fields to grant access to that are Subscriptions (leave blank for all)
+   */
+  public grantSubscription(grantee: IGrantable, fields?: [string]): Grant {
+    return this.grantType(grantee, 'Subscription', fields);
+  }
+
+  /**
+   * Adds an IAM policy statement for Type access to this GraphQLApi to an IAM
+   * principal's policy.
+   *
+   * @param grantee The principal (no-op if undefined)
+   * @param type The type to grant access
+   * @param fields The fields to grant access to that are of @param type (leave blank for all)
+   */
+  public grantType(grantee: IGrantable, type: string, fields?: [string]): Grant {
+    return Grant.addToPrincipal({
+      grantee,
+      actions: ['appsync:graphql'],
+      resourceArns: fields ?
```

Comment:
Same with this.. there is probably a better way to do this...

maybe like that a function that does a singular instance of the resourceArn

```
private genResourceArn(resource: IamResource): string {
  var name = resource.type ? resource.field ? `types/${resource.type}/fields/${field}`: `types/${resource.type}/*` :  '*'; 
  name = resource.custom ? resource.custom : name;
  return Stack.of(this).formatArn({
     service: 'appsync',
     resource: `apis/${this.apiId}`,
     sep: '/',
     resourceName: `${name}`,
   });
}
```

wouldn't solve the problem of empty fields i guess for Type... bu could be useful later